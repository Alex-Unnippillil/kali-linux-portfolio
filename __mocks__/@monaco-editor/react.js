import React from 'react';
export default function MonacoMock({ value = '', onChange }) {
  return React.createElement('textarea', {
    value,
    onChange: e => onChange && onChange(e.target.value),
  });
}
