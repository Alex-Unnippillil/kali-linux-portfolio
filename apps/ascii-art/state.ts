import usePersistentState from '../../hooks/usePersistentState';

export interface AsciiArtPreset {
  text: string;
  font: string;
  brightness: number;
  contrast: number;
}

export function useAsciiArtState() {
  const [presets, setPresets] = usePersistentState<Record<string, AsciiArtPreset>>(
    'ascii-art-presets',
    {}
  );
  const [activePreset, setActivePreset] = usePersistentState<string | null>(
    'ascii-art-active-preset',
    null
  );

  const savePreset = (name: string, preset: AsciiArtPreset) => {
    setPresets({ ...presets, [name]: preset });
    setActivePreset(name);
  };

  const loadPreset = (name: string) => {
    const preset = presets[name];
    if (preset) setActivePreset(name);
    return preset;
  };

  return { presets, savePreset, loadPreset, activePreset };
}
