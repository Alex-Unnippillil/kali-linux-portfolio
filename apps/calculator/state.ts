import usePersistentState from '../../hooks/usePersistentState';
import { profileSelector } from '../../utils/stateProfiler';

export const VARS_KEY = 'calc-vars';

export type VarMap = Record<string, string>;

export function useVariables() {
  return usePersistentState<VarMap>(
    VARS_KEY,
    () => ({}),
    (v): v is VarMap =>
      typeof v === 'object' && v !== null && !Array.isArray(v) &&
      Object.values(v).every((val) => typeof val === 'string'),
  );
}

export function loadVariables(): VarMap {
  return profileSelector(
    'calculator.loadVariables',
    () => {
      if (typeof window === 'undefined') return {};
      try {
        const stored = window.localStorage.getItem(VARS_KEY);
        if (stored) {
          const parsed = JSON.parse(stored);
          if (
            typeof parsed === 'object' &&
            parsed !== null &&
            !Array.isArray(parsed)
          ) {
            return parsed as VarMap;
          }
        }
      } catch {
        // ignore
      }
      return {};
    },
    { metadata: { store: 'localStorage', key: VARS_KEY } },
  );
}
