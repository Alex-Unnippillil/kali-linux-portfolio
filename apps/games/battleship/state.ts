import { BOARD_SIZE, SHIPS, Layout } from './ai';

export const LAYOUT_STORAGE_KEY = 'battleship:layout';

export type Orientation = 0 | 1;

export interface StoredShipPlacement {
  len: number;
  x: number;
  y: number;
  dir: Orientation;
}

export type StoredLayout = StoredShipPlacement[];

type StorageLike = Pick<Storage, 'getItem' | 'setItem' | 'removeItem'>;

const expectedLengths = (() => {
  const sorted = [...SHIPS].sort((a, b) => a - b);
  return sorted.join(',');
})();

function getStorage(storage?: StorageLike) {
  if (storage) return storage;
  if (typeof window === 'undefined') return undefined;
  try {
    return window.localStorage;
  } catch {
    return undefined;
  }
}

const ORIENTATION: Orientation[] = [0, 1];

const toInt = (value: unknown): number =>
  typeof value === 'number' ? value : Number.parseInt(String(value ?? 0), 10);

export function computeCells(ship: StoredShipPlacement): number[] {
  const cells: number[] = [];
  for (let i = 0; i < ship.len; i += 1) {
    const cx = ship.x + (ship.dir === 0 ? i : 0);
    const cy = ship.y + (ship.dir === 1 ? i : 0);
    if (cx < 0 || cy < 0 || cx >= BOARD_SIZE || cy >= BOARD_SIZE) {
      return [];
    }
    cells.push(cy * BOARD_SIZE + cx);
  }
  return cells;
}

function sanitizePlacement(ship: any): StoredShipPlacement {
  return {
    len: toInt(ship?.len),
    x: toInt(ship?.x),
    y: toInt(ship?.y),
    dir: ORIENTATION.includes(ship?.dir) ? (ship.dir as Orientation) : 0,
  };
}

function sameLengths(layout: StoredLayout) {
  const sorted = layout.map((ship) => ship.len).sort((a, b) => a - b);
  return sorted.join(',') === expectedLengths;
}

function hasOverlapOrAdjacency(
  cells: number[],
  occupied: Set<number>,
): boolean {
  for (const idx of cells) {
    if (occupied.has(idx)) return true;
    const cx = idx % BOARD_SIZE;
    const cy = Math.floor(idx / BOARD_SIZE);
    for (let dx = -1; dx <= 1; dx += 1) {
      for (let dy = -1; dy <= 1; dy += 1) {
        if (dx === 0 && dy === 0) continue;
        const nx = cx + dx;
        const ny = cy + dy;
        if (nx < 0 || ny < 0 || nx >= BOARD_SIZE || ny >= BOARD_SIZE) continue;
        const nIdx = ny * BOARD_SIZE + nx;
        if (occupied.has(nIdx)) return true;
      }
    }
  }
  return false;
}

export function isValidLayout(layout: unknown): layout is StoredLayout {
  if (!Array.isArray(layout) || layout.length !== SHIPS.length) {
    return false;
  }
  const placements = layout.map(sanitizePlacement);
  if (!sameLengths(placements)) return false;
  const occupied = new Set<number>();
  for (const ship of placements) {
    if (!Number.isFinite(ship.x) || !Number.isFinite(ship.y)) return false;
    if (!Number.isFinite(ship.len) || ship.len <= 0) return false;
    if (!ORIENTATION.includes(ship.dir)) return false;
    const cells = computeCells(ship);
    if (cells.length !== ship.len) return false;
    if (hasOverlapOrAdjacency(cells, occupied)) return false;
    cells.forEach((c) => occupied.add(c));
  }
  return true;
}

export function normalizeLayout(layout: unknown): StoredLayout | null {
  if (!Array.isArray(layout)) return null;
  const placements = layout.map(sanitizePlacement);
  return isValidLayout(placements) ? placements : null;
}

export function serializeShips(
  ships: Array<{
    len: number;
    x: number;
    y: number;
    dir: Orientation;
  }>,
): StoredLayout {
  return ships.map((ship) => ({
    len: ship.len,
    x: ship.x,
    y: ship.y,
    dir: ship.dir,
  }));
}

export function expandLayout(layout: StoredLayout): Layout[] {
  return layout.map((ship) => ({
    x: ship.x,
    y: ship.y,
    dir: ship.dir,
    len: ship.len,
    cells: computeCells(ship),
  }));
}

export function saveLayout(
  layout: StoredLayout,
  storage?: StorageLike,
): void {
  if (!isValidLayout(layout)) {
    throw new Error('Invalid layout');
  }
  const target = getStorage(storage);
  if (!target) return;
  try {
    target.setItem(LAYOUT_STORAGE_KEY, JSON.stringify(layout));
  } catch {
    // ignore storage failures
  }
}

export function loadLayout(storage?: StorageLike): StoredLayout | null {
  const target = getStorage(storage);
  if (!target) return null;
  try {
    const raw = target.getItem(LAYOUT_STORAGE_KEY);
    if (!raw) return null;
    const parsed = JSON.parse(raw);
    return normalizeLayout(parsed);
  } catch {
    return null;
  }
}

export function clearLayout(storage?: StorageLike): void {
  const target = getStorage(storage);
  if (!target) return;
  try {
    target.removeItem(LAYOUT_STORAGE_KEY);
  } catch {
    // ignore storage failures
  }
}
