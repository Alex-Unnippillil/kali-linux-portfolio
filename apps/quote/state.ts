import usePersistentState from '../../hooks/usePersistentState';

export function useQuoteNotification() {
  const [time, setTime] = usePersistentState<string>(
    'quote-notify-time',
    '09:00',
    (v): v is string => typeof v === 'string',
  );
  const [enabled, setEnabled] = usePersistentState<boolean>(
    'quote-notify-enabled',
    false,
    (v): v is boolean => typeof v === 'boolean',
  );
  return { time, setTime, enabled, setEnabled } as const;
}

