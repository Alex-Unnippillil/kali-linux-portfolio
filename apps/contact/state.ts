import { useCallback, useEffect, useState } from 'react';
import usePersistentState from '../../hooks/usePersistentState';
import { sendEmail } from './utils/sendEmail';

interface Draft {
  name: string;
  email: string;
  message: string;
}

type Status = 'idle' | 'queued' | 'sending' | 'sent' | 'error';

export const useContactState = () => {
  const [draft, setDraft, resetDraft] = usePersistentState<Draft>(
    'contact-draft',
    { name: '', email: '', message: '' },
  );
  const [status, setStatus] = useState<Status>('idle');
  const [queued, setQueued] = useState(false);

  const attemptSend = useCallback(async () => {
    try {
      setStatus('sending');
      await sendEmail(draft);
      setStatus('sent');
      resetDraft();
    } catch {
      setStatus('error');
    }
  }, [draft, resetDraft]);

  const send = useCallback(() => {
    if (!navigator.onLine) {
      setStatus('queued');
      setQueued(true);
      return;
    }
    void attemptSend();
  }, [attemptSend]);

  useEffect(() => {
    if (!queued) return;
    const handleOnline = () => {
      setQueued(false);
      void attemptSend();
    };
    window.addEventListener('online', handleOnline);
    return () => window.removeEventListener('online', handleOnline);
  }, [queued, attemptSend]);

  return {
    draft,
    setDraft,
    send,
    status,
    reset: resetDraft,
  } as const;
};

export type ContactState = ReturnType<typeof useContactState>;

