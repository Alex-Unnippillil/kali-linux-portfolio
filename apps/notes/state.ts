"use client";

import { useCallback, useEffect, useMemo } from 'react';
import usePersistentState from '../../hooks/usePersistentState';

export interface NoteLabel {
  id: string;
  name: string;
  color: string;
}

export interface Note {
  id: string;
  title: string;
  content: string;
  labelIds: string[];
  archived: boolean;
  order: number;
  createdAt: string;
  updatedAt: string;
}

export interface NotesFilter {
  labelIds: string[];
  /**
   * When true, the UI should display archived notes instead of active ones.
   */
  showArchived: boolean;
}

export interface NotesState {
  notes: Note[];
  labels: NoteLabel[];
  filter: NotesFilter;
}

type NotesStateStorage = {
  notes?: Partial<Note>[];
  labels?: Partial<NoteLabel>[];
  filter?: Partial<NotesFilter>;
};

const STORAGE_KEY = 'notes-app-state-v1';

const DEFAULT_LABELS: readonly NoteLabel[] = [
  { id: 'inbox', name: 'Inbox', color: '#3B82F6' },
  { id: 'work', name: 'Work', color: '#F97316' },
  { id: 'research', name: 'Research', color: '#8B5CF6' },
  { id: 'personal', name: 'Personal', color: '#10B981' },
];

function createDefaultLabels(): NoteLabel[] {
  return DEFAULT_LABELS.map((label) => ({ ...label }));
}

function createDefaultState(): NotesState {
  return {
    notes: [],
    labels: createDefaultLabels(),
    filter: { labelIds: [], showArchived: false },
  };
}

function isValidStorage(value: unknown): value is NotesStateStorage {
  if (!value || typeof value !== 'object') return false;
  const storage = value as NotesStateStorage;
  if ('notes' in storage && !Array.isArray(storage.notes)) return false;
  if ('labels' in storage && !Array.isArray(storage.labels)) return false;
  if ('filter' in storage && typeof storage.filter !== 'object') return false;
  return true;
}

function sanitizeLabel(partial: Partial<NoteLabel> | undefined, index: number): NoteLabel | null {
  if (!partial || typeof partial !== 'object') return null;
  const fallback = DEFAULT_LABELS[index % DEFAULT_LABELS.length];
  const idCandidate =
    typeof partial.id === 'string' && partial.id.trim()
      ? partial.id.trim()
      : fallback?.id ?? `label-${index + 1}`;
  const name =
    typeof partial.name === 'string' && partial.name.trim()
      ? partial.name.trim()
      : fallback?.name ?? `Label ${index + 1}`;
  const color =
    typeof partial.color === 'string' && partial.color.trim()
      ? partial.color.trim()
      : fallback?.color ?? '#3B82F6';
  return { id: idCandidate, name, color };
}

function sanitizeLabels(raw: Partial<NoteLabel>[] | undefined): NoteLabel[] {
  const seen = new Set<string>();
  const source = Array.isArray(raw) && raw.length > 0 ? raw : createDefaultLabels();
  const labels: NoteLabel[] = [];
  source.forEach((partial, index) => {
    const candidate = sanitizeLabel(partial, index);
    if (!candidate) return;
    let id = candidate.id;
    let counter = 1;
    while (seen.has(id)) {
      id = `${candidate.id}-${counter++}`;
    }
    seen.add(id);
    labels.push({ ...candidate, id });
  });
  return labels.length > 0 ? labels : createDefaultLabels();
}

function sanitizeNote(
  partial: Partial<Note> | undefined,
  index: number,
  validLabelIds: Set<string>,
): Note | null {
  if (!partial || typeof partial !== 'object') return null;
  const now = new Date().toISOString();
  const id =
    typeof partial.id === 'string' && partial.id.trim()
      ? partial.id.trim()
      : typeof partial.id === 'number'
      ? `note-${partial.id}`
      : `note-${index + 1}`;
  const title = typeof partial.title === 'string' ? partial.title : '';
  const content = typeof partial.content === 'string' ? partial.content : '';
  const archived = typeof partial.archived === 'boolean' ? partial.archived : false;
  const order =
    typeof partial.order === 'number' && Number.isFinite(partial.order)
      ? partial.order
      : index;
  const createdAt =
    typeof partial.createdAt === 'string' && partial.createdAt
      ? partial.createdAt
      : now;
  const updatedAt =
    typeof partial.updatedAt === 'string' && partial.updatedAt
      ? partial.updatedAt
      : createdAt;
  const labelIds = Array.isArray(partial.labelIds)
    ? Array.from(
        new Set(
          partial.labelIds.filter(
            (labelId): labelId is string =>
              typeof labelId === 'string' && validLabelIds.has(labelId),
          ),
        ),
      )
    : [];

  return {
    id,
    title,
    content,
    labelIds,
    archived,
    order,
    createdAt,
    updatedAt,
  };
}

function sanitizeNotes(
  raw: Partial<Note>[] | undefined,
  validLabelIds: Set<string>,
): Note[] {
  if (!Array.isArray(raw)) return [];
  const notes: Note[] = [];
  const seen = new Set<string>();
  raw.forEach((partial, index) => {
    const note = sanitizeNote(partial, index, validLabelIds);
    if (!note) return;
    let id = note.id;
    let counter = 1;
    while (seen.has(id)) {
      id = `${note.id}-${counter++}`;
    }
    seen.add(id);
    notes.push({ ...note, id });
  });
  notes.sort((a, b) => {
    if (a.order !== b.order) return a.order - b.order;
    if (a.createdAt !== b.createdAt) return a.createdAt.localeCompare(b.createdAt);
    return a.id.localeCompare(b.id);
  });
  return notes.map((note, index) =>
    note.order === index ? note : { ...note, order: index },
  );
}

function sanitizeFilter(
  partial: Partial<NotesFilter> | undefined,
  validLabelIds: Set<string>,
): NotesFilter {
  if (!partial || typeof partial !== 'object') {
    return { labelIds: [], showArchived: false };
  }
  const labelIds = Array.isArray(partial.labelIds)
    ? partial.labelIds.filter((id): id is string => typeof id === 'string' && validLabelIds.has(id))
    : [];
  const unique = Array.from(new Set(labelIds));
  const showArchived = typeof partial.showArchived === 'boolean' ? partial.showArchived : false;
  return { labelIds: unique, showArchived };
}

export function sanitizeNotesState(value: unknown): NotesState {
  const storage: NotesStateStorage = isValidStorage(value) ? value : {};
  const labels = sanitizeLabels(storage.labels);
  const labelIdSet = new Set(labels.map((label) => label.id));
  const notes = sanitizeNotes(storage.notes, labelIdSet);
  const filter = sanitizeFilter(storage.filter, labelIdSet);

  const normalizedNotes = notes.map((note) => ({
    ...note,
    labelIds: note.labelIds.filter((id) => labelIdSet.has(id)),
  }));

  return {
    notes: normalizedNotes,
    labels,
    filter,
  };
}

function statesEqual(a: unknown, b: unknown): boolean {
  try {
    return JSON.stringify(a) === JSON.stringify(b);
  } catch {
    return false;
  }
}

function generateId() {
  if (typeof crypto !== 'undefined' && 'randomUUID' in crypto) {
    try {
      return crypto.randomUUID();
    } catch {
      // fall back if randomUUID is not available
    }
  }
  return `note-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`;
}

export function reorderNotesArray(
  notes: Note[],
  noteId: string,
  targetId: string | null,
): Note[] {
  const ordered = [...notes].sort((a, b) => a.order - b.order);
  const sourceIndex = ordered.findIndex((note) => note.id === noteId);
  if (sourceIndex === -1) return ordered;
  const [moved] = ordered.splice(sourceIndex, 1);
  let targetIndex = targetId ? ordered.findIndex((note) => note.id === targetId) : -1;
  if (targetIndex < 0 || targetIndex > ordered.length) {
    targetIndex = ordered.length;
  }
  ordered.splice(targetIndex, 0, moved);
  const now = new Date().toISOString();
  return ordered.map((note, index) => {
    if (note.order === index) return note;
    if (note.id === moved.id) {
      return { ...note, order: index, updatedAt: now };
    }
    return { ...note, order: index };
  });
}

export function useNotesState() {
  const [storedState, setStoredState, resetStoredState] = usePersistentState<NotesStateStorage>(
    STORAGE_KEY,
    () => createDefaultState(),
    isValidStorage,
  );

  const state = useMemo(() => sanitizeNotesState(storedState), [storedState]);

  useEffect(() => {
    if (!statesEqual(storedState, state)) {
      setStoredState(state);
    }
  }, [state, storedState, setStoredState]);

  const applyStateUpdate = useCallback(
    (updater: (previous: NotesState) => NotesState) => {
      setStoredState((prev) => {
        const base = sanitizeNotesState(prev);
        const next = sanitizeNotesState(updater(base));
        return next;
      });
    },
    [setStoredState],
  );

  const addNote = useCallback(
    (defaults?: Partial<Pick<Note, 'title' | 'content' | 'labelIds'>>) => {
      applyStateUpdate((prev) => {
        const validLabelIds = new Set(prev.labels.map((label) => label.id));
        let initialLabels: string[] = [];
        if (Array.isArray(defaults?.labelIds)) {
          initialLabels = defaults.labelIds.filter((id) => validLabelIds.has(id));
        } else if (prev.filter.labelIds.length === 1) {
          initialLabels = prev.filter.labelIds;
        }
        const now = new Date().toISOString();
        const note: Note = {
          id: generateId(),
          title: typeof defaults?.title === 'string' ? defaults.title : '',
          content: typeof defaults?.content === 'string' ? defaults.content : '',
          labelIds: Array.from(new Set(initialLabels)),
          archived: false,
          order: prev.notes.length,
          createdAt: now,
          updatedAt: now,
        };
        return { ...prev, notes: [...prev.notes, note] };
      });
    },
    [applyStateUpdate],
  );

  const updateNote = useCallback(
    (noteId: string, changes: Partial<Pick<Note, 'title' | 'content' | 'labelIds'>>) => {
      applyStateUpdate((prev) => {
        const index = prev.notes.findIndex((note) => note.id === noteId);
        if (index === -1) return prev;
        const validLabelIds = new Set(prev.labels.map((label) => label.id));
        const nextNotes = [...prev.notes];
        const current = nextNotes[index];
        let modified = false;
        let nextNote = { ...current };
        if (typeof changes.title === 'string' && changes.title !== current.title) {
          nextNote = { ...nextNote, title: changes.title, updatedAt: new Date().toISOString() };
          modified = true;
        }
        if (typeof changes.content === 'string' && changes.content !== current.content) {
          nextNote = { ...nextNote, content: changes.content, updatedAt: new Date().toISOString() };
          modified = true;
        }
        if (Array.isArray(changes.labelIds)) {
          const sanitizedLabels = Array.from(
            new Set(changes.labelIds.filter((id) => typeof id === 'string' && validLabelIds.has(id))),
          );
          if (JSON.stringify(sanitizedLabels) !== JSON.stringify(current.labelIds)) {
            nextNote = {
              ...nextNote,
              labelIds: sanitizedLabels,
              updatedAt: new Date().toISOString(),
            };
            modified = true;
          }
        }
        if (!modified) return prev;
        nextNotes[index] = nextNote;
        return { ...prev, notes: nextNotes };
      });
    },
    [applyStateUpdate],
  );

  const deleteNote = useCallback(
    (noteId: string) => {
      applyStateUpdate((prev) => {
        if (!prev.notes.some((note) => note.id === noteId)) return prev;
        const remaining = prev.notes
          .filter((note) => note.id !== noteId)
          .map((note, index) => (note.order === index ? note : { ...note, order: index }));
        return { ...prev, notes: remaining };
      });
    },
    [applyStateUpdate],
  );

  const toggleArchive = useCallback(
    (noteId: string) => {
      applyStateUpdate((prev) => {
        const now = new Date().toISOString();
        const notes = prev.notes.map((note) =>
          note.id === noteId
            ? { ...note, archived: !note.archived, updatedAt: now }
            : note,
        );
        return { ...prev, notes };
      });
    },
    [applyStateUpdate],
  );

  const toggleNoteLabel = useCallback(
    (noteId: string, labelId: string) => {
      applyStateUpdate((prev) => {
        const valid = prev.labels.some((label) => label.id === labelId);
        if (!valid) return prev;
        const now = new Date().toISOString();
        const notes = prev.notes.map((note) => {
          if (note.id !== noteId) return note;
          const hasLabel = note.labelIds.includes(labelId);
          const nextLabels = hasLabel
            ? note.labelIds.filter((id) => id !== labelId)
            : [...note.labelIds, labelId];
          return { ...note, labelIds: nextLabels, updatedAt: now };
        });
        return { ...prev, notes };
      });
    },
    [applyStateUpdate],
  );

  const moveNote = useCallback(
    (noteId: string, targetId: string | null) => {
      applyStateUpdate((prev) => {
        const reordered = reorderNotesArray(prev.notes, noteId, targetId);
        if (statesEqual(reordered, prev.notes)) return prev;
        return { ...prev, notes: reordered };
      });
    },
    [applyStateUpdate],
  );

  const setFilterLabel = useCallback(
    (labelId: string) => {
      applyStateUpdate((prev) => {
        if (!prev.labels.some((label) => label.id === labelId)) return prev;
        const current = prev.filter.labelIds;
        const set = new Set(current);
        if (set.has(labelId)) {
          set.delete(labelId);
        } else {
          set.add(labelId);
        }
        const orderMap = new Map(prev.labels.map((label, index) => [label.id, index] as const));
        const nextLabels = Array.from(set).sort((a, b) =>
          (orderMap.get(a) ?? 0) - (orderMap.get(b) ?? 0),
        );
        if (JSON.stringify(nextLabels) === JSON.stringify(current)) return prev;
        return { ...prev, filter: { ...prev.filter, labelIds: nextLabels } };
      });
    },
    [applyStateUpdate],
  );

  const clearLabelFilters = useCallback(() => {
    applyStateUpdate((prev) => {
      if (prev.filter.labelIds.length === 0) return prev;
      return { ...prev, filter: { ...prev.filter, labelIds: [] } };
    });
  }, [applyStateUpdate]);

  const setShowArchived = useCallback(
    (show: boolean) => {
      applyStateUpdate((prev) => {
        if (prev.filter.showArchived === show) return prev;
        return { ...prev, filter: { ...prev.filter, showArchived: show } };
      });
    },
    [applyStateUpdate],
  );

  const reset = useCallback(() => {
    resetStoredState();
  }, [resetStoredState]);

  return {
    state,
    actions: {
      addNote,
      updateNote,
      deleteNote,
      toggleArchive,
      toggleNoteLabel,
      moveNote,
      setFilterLabel,
      clearLabelFilters,
      setShowArchived,
    },
    reset,
  } as const;
}

