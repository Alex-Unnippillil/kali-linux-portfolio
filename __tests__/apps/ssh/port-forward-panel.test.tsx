import React from 'react';
import { fireEvent, render, screen, waitFor, within } from '@testing-library/react';
import PortForwardPanel from '@/apps/ssh/components/PortForwardPanel';
import { __testing as profileTesting } from '@/apps/ssh/state/profiles';

describe('PortForwardPanel', () => {
  beforeEach(() => {
    window.localStorage.clear();
    profileTesting.reset();
  });

  it('matches the snapshot for an empty profile', () => {
    const { asFragment } = render(<PortForwardPanel profileId="snapshot-profile" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('updates status chip when activating a forward', async () => {
    render(<PortForwardPanel profileId="profile-1" />);

    fireEvent.change(screen.getByLabelText('Source port'), { target: { value: '3000' } });
    fireEvent.change(screen.getByLabelText('Destination host'), {
      target: { value: 'internal.service' },
    });
    fireEvent.change(screen.getByLabelText('Destination port'), { target: { value: '8080' } });
    fireEvent.click(screen.getByRole('button', { name: 'Add Forward' }));

    const item = screen.getByRole('listitem');
    expect(within(item).getByText('Stopped')).toBeInTheDocument();

    const toggle = within(item).getByLabelText('Activate forward 3000 to internal.service:8080');
    fireEvent.click(toggle);

    await waitFor(() => expect(toggle).toBeChecked());
    await waitFor(() => expect(within(item).getByText('Listening')).toBeInTheDocument());

    const events = profileTesting.getEvents();
    expect(events).toEqual([
      expect.objectContaining({ type: 'started', profileId: 'profile-1' }),
    ]);
  });

  it('tears down active listeners on unmount', async () => {
    const { unmount } = render(<PortForwardPanel profileId="cleanup" />);

    fireEvent.change(screen.getByLabelText('Source port'), { target: { value: '4000' } });
    fireEvent.change(screen.getByLabelText('Destination host'), {
      target: { value: 'db.internal' },
    });
    fireEvent.change(screen.getByLabelText('Destination port'), { target: { value: '5432' } });
    fireEvent.click(screen.getByRole('button', { name: 'Add Forward' }));

    const item = screen.getByRole('listitem');
    const toggle = within(item).getByLabelText('Activate forward 4000 to db.internal:5432');
    fireEvent.click(toggle);

    await waitFor(() => expect(toggle).toBeChecked());

    unmount();

    const events = profileTesting.getEvents();
    expect(events).toEqual([
      expect.objectContaining({ type: 'started', profileId: 'cleanup' }),
      expect.objectContaining({ type: 'stopped', profileId: 'cleanup' }),
    ]);
  });
});

