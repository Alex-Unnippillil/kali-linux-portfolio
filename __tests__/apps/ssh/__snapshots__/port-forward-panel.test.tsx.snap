// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PortForwardPanel matches the snapshot for an empty profile 1`] = `
<DocumentFragment>
  <section
    class="mt-6 rounded border border-gray-700 bg-gray-900 p-4"
  >
    <div
      class="mb-4 flex flex-col gap-2 md:flex-row md:items-center md:justify-between"
    >
      <div>
        <h2
          class="text-lg font-semibold text-white"
        >
          Port Forwarding
        </h2>
        <p
          class="text-sm text-gray-300"
        >
          Define local and remote forwards. This panel simulates SSH listeners and does not open real sockets.
        </p>
      </div>
    </div>
    <form
      class="mb-4 grid gap-4 md:grid-cols-[auto,1fr,1fr,auto] md:items-end"
    >
      <div
        class="flex flex-col"
      >
        <span
          class="mb-1 text-xs font-semibold uppercase tracking-wide text-gray-400"
        >
          Type
        </span>
        <button
          aria-label="Toggle forward direction"
          class="rounded border border-gray-600 bg-gray-800 px-3 py-2 text-sm text-white hover:border-gray-400"
          type="button"
        >
          Local → Remote
        </button>
      </div>
      <label
        class="flex flex-col"
        for="«r0»-source-port"
      >
        <span
          class="mb-1 text-xs font-semibold uppercase tracking-wide text-gray-400"
        >
          Source port
        </span>
        <input
          class="rounded border border-gray-600 bg-gray-800 p-2 text-white focus:border-blue-500 focus:outline-none"
          id="«r0»-source-port"
          min="1"
          type="number"
          value="8080"
        />
      </label>
      <label
        class="flex flex-col"
        for="«r0»-destination-host"
      >
        <span
          class="mb-1 text-xs font-semibold uppercase tracking-wide text-gray-400"
        >
          Destination host
        </span>
        <input
          class="rounded border border-gray-600 bg-gray-800 p-2 text-white focus:border-blue-500 focus:outline-none"
          id="«r0»-destination-host"
          type="text"
          value="127.0.0.1"
        />
      </label>
      <label
        class="flex flex-col"
        for="«r0»-destination-port"
      >
        <span
          class="mb-1 text-xs font-semibold uppercase tracking-wide text-gray-400"
        >
          Destination port
        </span>
        <div
          class="flex gap-2"
        >
          <input
            class="w-full rounded border border-gray-600 bg-gray-800 p-2 text-white focus:border-blue-500 focus:outline-none"
            id="«r0»-destination-port"
            min="1"
            type="number"
            value="80"
          />
          <button
            class="whitespace-nowrap rounded bg-blue-600 px-3 py-2 text-sm font-semibold text-white hover:bg-blue-500"
            type="submit"
          >
            Add Forward
          </button>
        </div>
      </label>
    </form>
    <p
      class="text-sm text-gray-400"
    >
      No forwards defined yet. Add one above to start listening.
    </p>
  </section>
</DocumentFragment>
`;
